File: qc_lab_simulator/flashcards/markup.py
"""Minimal safe markup renderer shared by HTML preview and Anki CSV export."""

from __future__ import annotations

import html
import re

from .localization import semantic_label

SEMANTIC_KINDS = {"rule", "warning", "rejection", "specimen", "instrument", "qc"}

_SEMANTIC_PATTERN = re.compile(r"\[\[(?P<kind>[a-z]+):(?P<text>[^\]]+)\]\]")
_CODE_PATTERN = re.compile(r"`([^`\n]+)`")
_BOLD_PATTERN = re.compile(r"\*\*([^*\n]+)\*\*")
_ITALIC_PATTERN = re.compile(r"(?<!\*)\*([^*\n]+)\*(?!\*)")


def validate_markup_text(text: str) -> list[str]:
    errors: list[str] = []
    if text.count("**") % 2 != 0:
        errors.append("Unbalanced bold marker '**'.")
    italic_markers = re.findall(r"(?<!\*)\*(?!\*)", text)
    if len(italic_markers) % 2 != 0:
        errors.append("Unbalanced italic marker '*'.")
    if text.count("`") % 2 != 0:
        errors.append("Unbalanced monospace marker '`'.")

    if text.count("[[") != text.count("]]"):
        errors.append("Unbalanced semantic token marker '[[' or ']]'.")

    for match in _SEMANTIC_PATTERN.finditer(text):
        kind = match.group("kind")
        token_text = match.group("text").strip()
        if kind not in SEMANTIC_KINDS:
            errors.append(f"Unsupported semantic token '{kind}'.")
        if not token_text:
            errors.append(f"Semantic token '{kind}' cannot be empty.")
    return errors


def render_markup_to_html(text: str, *, language: str = "es") -> str:
    escaped = html.escape(text)

    def semantic_repl(match: re.Match[str]) -> str:
        kind = match.group("kind")
        token_text = html.escape(match.group("text").strip())
        if kind not in SEMANTIC_KINDS:
            return html.escape(match.group(0))
        label = html.escape(semantic_label(kind, language))
        return (
            f'<span class="semantic semantic-{kind}">'
            f'<span class="semantic-label">{label}</span> {token_text}'
            f"</span>"
        )

    rendered = _SEMANTIC_PATTERN.sub(semantic_repl, escaped)
    rendered = _CODE_PATTERN.sub(r"<code>\1</code>", rendered)
    rendered = _BOLD_PATTERN.sub(r"<strong>\1</strong>", rendered)
    rendered = _ITALIC_PATTERN.sub(r"<em>\1</em>", rendered)
    rendered = rendered.replace("\n", "<br>\n")
    return rendered

File: apps/student-next/src/markup.js
// src/markup.js
// Renders exactly the four canonical constructs from
// qc_lab_simulator/flashcards/markup.py: [[kind:text]], **bold**, *italic*,
// `code`. This is a renderer, not a new content model — do not add syntax.

const SEMANTIC_KINDS = new Set(["rule", "warning", "rejection", "specimen", "instrument", "qc"]);
const SEMANTIC_LABEL_ES = {
  rule: "Regla",
  warning: "Advertencia",
  rejection: "Rechazo",
  specimen: "Muestra",
  instrument: "Instrumento",
  qc: "Control de calidad",
};

const SEMANTIC_PATTERN = /\[\[([a-z]+):([^\]]+)\]\]/g;
const CODE_PATTERN = /`([^`\n]+)`/g;
const BOLD_PATTERN = /\*\*([^*\n]+)\*\*/g;
const ITALIC_PATTERN = /(?<!\*)\*([^*\n]+)\*(?!\*)/g;

function escapeHtml(text) {
  return String(text)
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;")
    .replaceAll("'", "&#39;");
}

export function renderCardMarkup(text) {
  let escaped = escapeHtml(text);

  escaped = escaped.replace(SEMANTIC_PATTERN, (match, kind, body) => {
    const trimmed = body.trim();
    if (!SEMANTIC_KINDS.has(kind)) return match;
    const label = SEMANTIC_LABEL_ES[kind] || kind;
    return `<span class="semantic semantic-${kind}"><span class="semantic-label">${label}</span> ${trimmed}</span>`;
  });

  escaped = escaped.replace(CODE_PATTERN, "<code>$1</code>");
  escaped = escaped.replace(BOLD_PATTERN, "<strong>$1</strong>");
  escaped = escaped.replace(ITALIC_PATTERN, "<em>$1</em>");
  escaped = escaped.replaceAll("\n", "<br>\n");

  return escaped;
}
